patches/server/0839-Freeze-Override-API.patch

```diff
@@ -0,0 +1,112 @@
++            this.entityData.set(Entity.DATA_TICKS_FROZEN, 0);
++        } else {
++            this.entityData.set(Entity.DATA_TICKS_FROZEN, realFrozenTicks);
++        }
++    }
++    // Paper end - Frozen Override API
+ 
+     public void setTicksFrozen(int frozenTicks) {
+-        this.entityData.set(Entity.DATA_TICKS_FROZEN, frozenTicks);
++        updateFrozenEffect(); // Paper - Frozen Override API
++        this.realFrozenTicks = frozenTicks; // Paper - Frozen Override API
+     }
+ 
+     public float getPercentFrozen() {
+@@ -2965,6 +2983,8 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
+     }
+ 
+     public boolean isFullyFrozen() {
++        if (this.frozenOverride == net.kyori.adventure.util.TriState.TRUE) {return true;} // Paper - Frozen Override API
++         else if (this.frozenOverride == net.kyori.adventure.util.TriState.FALSE) {return false;} // Paper - Frozen Override API
+         return this.getTicksFrozen() >= this.getTicksRequiredToFreeze();
+     }
+ 
+diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
+index fe7e22d9a69d69dfcce63faa28e90945ea45fc49..b1821dfa27a219d82e8720104891c5131cef6966 100644
+--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
++++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
+@@ -568,7 +568,7 @@ public abstract class LivingEntity extends Entity {
+ 
+     protected void tryAddFrost() {
+         if (!this.getBlockStateOn().isAir()) {
+-            int i = this.getTicksFrozen();
++            int i = this.getSentTicksFrozen(); // Paper - Frozen Override API
+ 
+             if (i > 0) {
+                 AttributeInstance attributemodifiable = this.getAttribute(Attributes.MOVEMENT_SPEED);
+@@ -577,7 +577,11 @@ public abstract class LivingEntity extends Entity {
+                     return;
+                 }
+ 
+-                float f = -0.05F * this.getPercentFrozen();
++                // Paper start - Frozen Override API
++                // See this.getPercentFrozen();
++                int requiredFreeze = this.getTicksRequiredToFreeze();
++                float f = -0.05F * ((float) Math.min(this.getSentTicksFrozen(), requiredFreeze) / (float) requiredFreeze);
++                // Paper end - Frozen Override API
+ 
+                 attributemodifiable.addTransientModifier(new AttributeModifier(LivingEntity.SPEED_MODIFIER_POWDER_SNOW_UUID, "Powder snow slow", (double) f, AttributeModifier.Operation.ADDITION));
+             }
+@@ -4276,6 +4280,9 @@ public abstract class LivingEntity extends Entity {
+ 
+     @Override
+     public boolean canFreeze() {
++        if (this.frozenOverride == net.kyori.adventure.util.TriState.TRUE) {return true;} // Paper - Frozen Override API
++        else if (this.frozenOverride == net.kyori.adventure.util.TriState.FALSE) {return false;} // Paper - Frozen Override API
```

Comment:
I'm not sure about this. I think the `canFreeze` method should be left to its vanilla logic, but rather in places where canFreeze is used, check the override. Or maybe make another method that checks it and call that one in place of canFreeze